### PR TITLE
Improve DD4hep workflow perf, step 3: Also reduced the number of paths (with regex) looped over for HCAL

### PR DESCRIPTION
**Pull request by ghugo83**
This is based on the exact same observation as in https://github.com/cms-sw/cmssw/pull/32505 : when SpecPar blocks have the same name, all their paths are stacked together. This results in irrelevant paths (with regex) being looped over, hence a performance penalty.

https://github.com/cms-sw/cmssw/pull/32505 proposed a simple fix for the Tracker (just addressing the issue directly in the XMLs); this is the equivalent for HCAL.
Unnecessary HCAL paths (with regex) were looped over, for example at https://github.com/cms-sw/cmssw/blob/master/SimG4Core/Geometry/src/DD4hep_DDG4Builder.cc#L62

**Summary of performance gains for FULL step 1 with DD4hep workflow (entire step with all initializations + XML parsing + CMS geo construction, up to event generation).**
ZMM, Run3, on my local (obviously only the order of magnitude is relevant).
- **https://github.com/cms-sw/cmssw/pull/32505: ~205s -> ~140s**
- **https://github.com/cms-sw/cmssw/pull/32511: ~140s -> ~85s**
- **This PR: ~85s -> ~80s**
- DDD is at ~42 s

NB: The last occurrence of this issue (for Run 3) is in ZDC (zdcsens.xml and zdcProdCuts.xml). I will not address the ZDC case, as the paths there have no regex, hence the perf penalty is completely negligible, not worth the XMLs files versioning, risk of regression, etc.

NB 2: https://github.com/cms-sw/cmssw/pull/32511 should be merged first. This branch is on top of my working branch, hence the diff with master will become relevant as soon as #32511 is merged.

@ianna @civanch @cvuosalo

### CODE REVIEW

File: Geometry/HcalCommonData/data/v1/hcalsenspmf.xml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0"?>
```
Comment (cvuosalo) [inline Geometry/HcalCommonData/data/v1/hcalsenspmf.xml:1]: Full path is wrong. It should be `Geometry/HcalCommonData/data/hcalsenspmf/v1/hcalsenspmf.xml`.

Comment (ghugo83) [reply Geometry/HcalCommonData/data/v1/hcalsenspmf.xml:1]: The previous path was Geometry/HcalCommonData/data/hcalsenspmf.xml

File: Geometry/HcalSimData/data/v1/HcalProdCuts.xml [hunk 2]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0"?>
```
Comment (cvuosalo) [inline Geometry/HcalSimData/data/v1/HcalProdCuts.xml:1]: Full path needs to be ` Geometry/HcalSimData/data/HcalProdCuts/2021/v2/HcalProdCuts.xml`. Make sure you derive `v2` from the `v1` file.

File: Geometry/CMSCommonData/data/dd4hep/cmsExtendedGeometry2021ZeroMaterial.xml [hunk 3]
```diff
@@ -355,7 +355,7 @@
     <Include ref='Geometry/VeryForwardData/data/CTPPS_Pixel_Sens.xml'/>
     <Include ref='Geometry/VeryForwardData/data/CTPPS_2018/RP_Dist_Beam_Cent.xml'/>
     <Include ref='Geometry/EcalSimData/data/ecalsens.xml'/>
-    <Include ref='Geometry/HcalCommonData/data/hcalsenspmf.xml'/>
```
Comment (ghugo83) [inline Geometry/CMSCommonData/data/dd4hep/cmsExtendedGeometry2021ZeroMaterial.xml:358]: here

File: Geometry/HcalCommonData/data/v1/hcalsenspmf.xml (hunk 1 above)
Comment (cvuosalo) [reply Geometry/HcalCommonData/data/v1/hcalsenspmf.xml:1]: Yes, the new directory `hcalsenspmf` has to be created.

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
@ghugo83 , likely there is a problem - see regression.

**Comment by ghugo83** [comment]
Alright, just pushed the v1 files renaming.

**Comment by ghugo83** [comment]
I had noticed that T2 does not contain the TrackingMaterialsGroup, this is an independent omission which was there before these 4 PRs: https://github.com/cms-sw/cmssw/blame/master/Configuration/Geometry/python/dict2021Geometry.py 
I would suggest it is addressed in an independent PR

**Comment by ghugo83** [comment]
The first 2 PRs provide most of the perf improvement.
This PR is related to HCAL XMLs and only provide ~5 s improvement instead, and points to look at regarding the HCAL results, hence I would suggest I look at the remaining points after the holiday, please do not merge this PR yet.

**Comment by ghugo83** [comment]
Alright, now am dead till January 6th :)

**Comment by ianna** [comment]
> Alright, now am dead till January 6th :)

Happy holidays!

**Comment by cvuosalo** [comment]
hold

**Comment by ghugo83** [comment]
> Happy holidays!

Thanks @ianna !
You too!

**Comment by civanch** [comment]
unhold

**Comment by ghugo83** [comment]
There is indeed diff in the results, but the error seems to not be in this PR, but in the master branch.
This PR slightly improves the perf, but also actually fixes what seems a major bug in all **DD4hep** workflows.

In https://github.com/cms-sw/cmssw/blob/master/DetectorDescription/DDCMS/plugins/dd4hep/DDDefinitions2Objects.cc#L974 :
the paths of different SpecPars are gathered.
This not only results in a performance penalty, but also leads to wrong results when the SpecPar define different paths: undesired paths will be considered for a given parameter.



**Example:**
Content of `HcalProdCuts.xml`:
```
<SpecPar name="hcal">
      <PartSelector path="//CALO"/>
      <PartSelector path="//MBAT"/>
      <PartSelector path="//MBBT.*"/>
      <PartSelector path="//VCAL"/>
      <Parameter name="CMSCutsRegion" value="HcalRegion" eval="false"/>
      <Parameter name="ProdCutsForElectrons" value="1*mm"/>
      <Parameter name="ProdCutsForPositrons" value="1*mm"/>
      <Parameter name="ProdCutsForGamma" value="1*mm"/>
    </SpecPar>
```

**In master:**
The paths defining the `CMSCutsRegion` are:
//HBS.*
//HTS.*
//HES.*
//HVQF
//HCal
//HBS.*
//HTS.*
//.*HES.*
//HVQX
//VcalElecCathode
//FibreBundle.*
//VcalFibreBundleL.*
//VcalFibreBundleR.*
//CALO
//MBAT
//MBBT.*
//VCAL
**This is not expected!!!**

**After this PR:**
The paths defining the CMSCutsRegion are:
//CALO
//MBAT
//MBBT.*
//VCAL
**This is what is expected.**

The reason why there are too many paths in master, associated to CMSCutsRegion, is because there are other, irrelevant, SpecPars with the same name `hcal`, defined in other XMLs:

-hcalSimNumbering.xml:
 ```
<SpecPar name="**hcal**">
    <PartSelector path="//HBS.*"/>
    <PartSelector path="//HTS.*"/>
    <PartSelector path="//HES.*"/>
    <PartSelector path="//HVQF"/>
    <Parameter name="OnlyForHcalSimNumbering" value="HCAL" eval="false"/>
  </SpecPar>
```
- hcalRecNumbering.xml:
```
 <SpecPar name="**hcal**">
    <PartSelector path="//HCal"/>
    <Parameter name="OnlyForHcalRecNumbering" value="HCAL" eval="false"/>
    <Parameter name="TopologyMode" value="HcalTopologyMode::SLHC" eval="false"/>
    <Parameter name="TriggerMode"  value="HcalTopologyMode::TriggerMode_2021" eval="false"/>
  </SpecPar>
```
- hcalsenspmf.xml:
```
<SpecPar name="**hcal**">
   <PartSelector path="//HBS.*"/>
   <PartSelector path="//HTS.*"/>
   <PartSelector path="//.*HES.*"/>
   <PartSelector path="//HVQX"/>
   <PartSelector path="//VcalElecCathode"/>
   <PartSelector path="//FibreBundle.*"/>
   <PartSelector path="//VcalFibreBundleL.*"/>
   <PartSelector path="//VcalFibreBundleR.*"/>
   <Parameter name="SensitiveDetector" value="HcalSensitiveDetector" eval="false"/>
   <Parameter name="ReadOutName" value="HcalHits" eval="false"/>
  </SpecPar>
```

As you can see, these files define parameters which are irrelevant for `CMSCutsRegion` (`OnlyForHcalSimNumbering`, `ReadOutName`), but their paths are still added to define CMSCutsRegion.
That is because the SpecPars all have the same name (`hcal`).

A similar issue appears in HCAL for all these parameters:
- OnlyForHcalSimNumbering
- OnlyForHcalRecNumbering
- TopologyMode
- TriggerMode
- SensitiveDetector
- ReadOutName
- CMSCutsRegion

This results in errors in HCAL SimHits in master branch (DD4hep workflows only): Energy deposits, EMfraction, sensitive detectors, time, depth.
Which in turns impact the steps downstream and the full workflow.

**Comment by ghugo83** [comment]
How to reproduce the issue in master, in DD4hep workflows:

```
const cms::DDDetector* det = compactView_->detector();
const dd4hep::SpecParRegistry& allSpecPars = det->specpars();
dd4hep::SpecParRefs filteredSpecPars;
allSpecPars.filter(filteredSpecPars, "OnlyForHcalSimNumbering");
for (auto const& specPar : filteredSpecPars) {
    std::cout << "SpecPar filtered with parameter OnlyForHcalSimNumbering, SpecPar name = " << specPar.first << std::endl;
    for (auto const &path : specPar.second->paths) {
      std::cout << path << std::endl;
    }
  }
```

In this example, `OnlyForHcalSimNumbering` parameter is used, but one can of course replace by any parameter of interest.
One can see that the list of paths associated to this parameters is not correct in master branch, and that it is correct after merge of this PR.

**Comment by ghugo83** [comment]
As mentioned in https://github.com/cms-sw/cmssw/pull/32505, in practice, there are 3 places where SpecPar have same names:
- Tracker
- HCAL
- ZDC
In Tracker, the SpecPars with same name define the same paths: hence there was no change in Tracker results (**only great performance improvement**, because the same paths were looped over twice). 

In HCAL instead, the SpecPars with same name define different paths: hence this PR has an impact on performance, and, unexpected outcome, also **affects the results.**

**Comment by ghugo83** [comment]
Plan of action:
- **Validation of HCAL SIM with DD4hep**: 
The fact that the results in master with DD4hep were not correct, implies that the validation oldDD-DD4hep for HCAL sim was not fully satisfactory, and hence should be re-done. 
Since the validation was not fully complete, it could be that there are still discrepancies oldDD versus DD4hep in HCAL sim.

- **Avoiding the same issue in DD4hep workflows to appear in other subdetectors / in Phase2 XMLs**:
The issue is fully fixed in this PR for Run 3 DD4hep workflow, for HCAL.
Though, I expect a similar issue in ZDC (not certain, depends how the code dealing with the parameters is implemented).
Plan:
- A minima, add a safety in https://github.com/cms-sw/cmssw/blob/master/DetectorDescription/DDCMS/plugins/dd4hep/DDDefinitions2Objects.cc#L974 , which throws an exception when the paths from 2 SpecPar with the same name but from different files are merged. 
This will allow to spot / fix any other occurrence of this bug.
- Possibly, only store the SpecPar name **with its namespace,** to directly avoid the issue (but this can not be done as-is, as there are harcoded SpecPar names (with no namespace) in the code... Hence this would need to be done very carefully (if ever possible), as it could lead to regressions.

**Comment by civanch** [comment]
@bsunanda , can you, please, comment on the problem.

@ghugo83 , thanks for this investigation. In my view, we cannot change names and xml for DDD but for DD4Hep only we may rename whatever. Also we can optimize any name inside Phase-2 detectors.

**Comment by bsunanda** [comment]
I understand the specpar names should be uniques throughout. I shall try to see this for other files relates to Hcal, HGCal and Ecal. I shall retest the equivalence of parameter loadings in these detectors after this PR is merged

**Comment by ghugo83** [comment]
@bsunanda 
Ok cool

**Comment by cvuosalo** [comment]
unhold

**Comment by cvuosalo** [comment]
@cmsbuild please test

**Comment by cvuosalo** [comment]
@ghugo83 If I understand correctly, the SpecPar name is never referenced, but it is used to group SpecPars together. That is, SpecPars with the same name in different files are all grouped together. Otherwise, the SpecPar name is not used. Is that correct?

File: Geometry/HcalSimData/data/HcalProdCuts/2021/v2/HcalProdCuts.xml [hunk 4]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0"?>
+<DDDefinition xmlns="http://www.cern.ch/cms/DDL" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.cern.ch/cms/DDL ../../../DetectorDescription/Schema/DDLSchema.xsd">
+  <SpecParSection label="regions.xml" eval="true">
+    <SpecPar name="hcalProdCutsPar">
+      <PartSelector path="//CALO"/>
+      <PartSelector path="//MBAT"/>
+      <PartSelector path="//MBBT.*"/>
+      <PartSelector path="//VCAL"/>
+      <Parameter name="CMSCutsRegion" value="HcalRegion" eval="false"/>
+      <Parameter name="ProdCutsForElectrons" value="1*mm"/>
```
Comment (cvuosalo) [inline Geometry/HcalSimData/data/HcalProdCuts/2021/v2/HcalProdCuts.xml:10]: Why was this line deleted?
```
<Parameter name="ProdCutsForProtons" value="1*mm"/>
```

Comment (cvuosalo) [reply Geometry/HcalSimData/data/HcalProdCuts/2021/v2/HcalProdCuts.xml:10]: Actually, that line is in "v1" but not the old `HcalProdCuts.xml`. I wonder why.

**Comment by cvuosalo** [comment]
What version of ` Geometry/TrackerRecoData/data/PhaseI/v1/trackerRecoMaterial.xml` is included in this PR? There is already a "v1" in the master.

**Comment by cvuosalo** [comment]
Why is `SimTracker/TrackerMaterialAnalysis/data/trackingMaterialGroups_ForPhaseI.xml ` included in this PR, along with `SimTracker/TrackerMaterialAnalysis/data/trackingMaterialGroups_ForPhaseI/v1/trackingMaterialGroups_ForPhaseI.xml`?

**Comment by ghugo83** [comment]
>@ghugo83 If I understand correctly, the SpecPar name is is used to group SpecPars together. That is, SpecPars with the same name in different files are all grouped together. 

Yes exactly

>Otherwise, the SpecPar name is not used. Is that correct?
No, in the codebase apparently, the SPecPar name is sometimes used to access a specific SpecPar directly.
Using DDFilteredView::get<std::vector<std::string>>, specPar(name), etc

File: Geometry/HcalSimData/data/HcalProdCuts/2021/v2/HcalProdCuts.xml (hunk 4 above)
Comment (ghugo83) [reply Geometry/HcalSimData/data/HcalProdCuts/2021/v2/HcalProdCuts.xml:10]: >Why was this line deleted?
<Parameter name="ProdCutsForProtons" value="1*mm"/>

Hmm that line was not deleted, the file which was used in 2021 scenario was `Geometry/HcalSimData/data/HcalProdCuts/HcalProdCuts.xml `, which does not contain it.
I noticed it indeed appear in v1 and not in the rest, but that v1 was unused.
This might or might not need to be addressed, but would say is independent from this PR, where we do not want to change the XMLs in use, but just rename the SpecPars (which... has an effect on the results :) ).

**Comment by ghugo83** [comment]
>What version of Geometry/TrackerRecoData/data/PhaseI/v1/trackerRecoMaterial.xml is included in this PR? There is already a "v1" in the master.
>Why is SimTracker/TrackerMaterialAnalysis/data/trackingMaterialGroups_ForPhaseI.xml included in this PR, along with SimTracker/TrackerMaterialAnalysis/data/trackingMaterialGroups_ForPhaseI/v1/trackingMaterialGroups_ForPhaseI.xml?

@cvuosalo Yes that's a weird feature of cms-git, which shows the diff of the PR branch with master **at the time the PR was created*, instead of **at the present time**.
The early commits in this PR are already included in master on other topics, but still appear in the diff with master, because master did not contain them at PR creation time.
I invite you to look at the real diff of this PR branch (ghugo83:dd4hep_perf_3) with present master: https://github.com/cms-sw/cmssw/compare/master...ghugo83:dd4hep_perf_3

As you can see, this PR does not touch trackerRecoMaterial.xml.
It also does not change any trackingMaterialGroups XML file, but it places them in a trackingMaterialGroups_ForPhaseI repo: data/v1/ -> data/trackingMaterialGroups_ForPhaseI/v1, as you suggested in a previous comment

File: Geometry/HcalSimData/data/HcalProdCuts/2021/v2/HcalProdCuts.xml (hunk 4 above)
Comment (cvuosalo) [reply Geometry/HcalSimData/data/HcalProdCuts/2021/v2/HcalProdCuts.xml:10]: @ghugo83 If you could follow up with this issue after this PR, that would be helpful. We should make sure the correct version of `HcalProdCuts.xml` is in use.

**Comment by cvuosalo** [comment]
> > @ghugo83 If I understand correctly, the SpecPar name is is used to group SpecPars together. That is, SpecPars with the same name in different files are all grouped together.
> 
> Yes exactly
> 
> > Otherwise, the SpecPar name is not used. Is that correct?
> 
> No, in the codebase apparently, the SpecPar name is sometimes used to access a specific SpecPar directly.
> Using `DDFilteredView::get<std::vector<std::string>>`, `specPar(name)`, etc

@ghugo83 It looks like in the case of the SpecPar names you changed, they are not used anywhere in the code, so it is safe to change the names.

**Comment by cvuosalo** [comment]
There are a lot of comparison test differences. There are differences in GEM quantities, which would not seem to have anything to do with the PR changes. Some of the differences may just be statistical fluctuations. @ghugo83 Can you comment?

**Comment by ghugo83** [comment]
>@ghugo83 It looks like in the case of the SpecPar names you changed, they are not used anywhere in the code, so it is safe to change the names.

Yes it practice, there were few use cases of "hcal' as a SpecPar name, but only for 1 XML (out of the 4 XMLs with same SpecPar names), so I have just renamed the 3 others.

**Comment by ghugo83** [comment]
>There are a lot of comparison test differences. There are differences in GEM quantities, which would not seem to have anything to do with the PR changes. Some of the differences may just be statistical fluctuations. @ghugo83 Can you comment?

Yes with 10 events the results themselves, in a sense, _are_ statistical fluctuations.
Though, they should point the dependencies of the PR diff to the results.
This PR should change the ECal SimHits.

It is not clear to me why GEM would be affected though: maybe because once SimHits in a certain area are modified, it can have an impact elsewhere?

**Comment by kpedro88** [comment]
Any changes in the SIM step will impact the random number sequence, which can cause fluctuations in any detector for any subsequent events.

**Comment by ghugo83** [comment]
>Any changes in the SIM step will impact the random number sequence, which can cause fluctuations in any detector for any subsequent events.

Ok thanks for confirming this

**Comment by cvuosalo** [comment]
I guess we can leave it to the validation process to check whether there is any real problem in the GEMs.

**Comment by cvuosalo** [comment]
+1

**Comment by civanch** [comment]
+1

the difference is seen in DD4Hep WF, so it is the bug fix for DD4Hep.

**Comment by srimanob** [comment]
+upgrade

To follow up after this PR:
- Keep eyes on validation report @chayanit

**Comment by qliphy** [comment]
+1
